core/src/bls_sigverify/sigverifier.rs

```diff
@@ -0,0 +1,1681 @@
+                            vote_message,
+                            bls_pubkey,
+                            pubkey,
+                        });
+                    } else {
+                        packet.meta_mut().set_discard(true);
+                    }
+                }
+                ConsensusMessage::Certificate(cert) => {
+                    // Only need certs newer than root slot
+                    if cert.cert_type.slot() <= root_bank.slot() {
+                        self.stats.received_old.fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    }
+
+                    if self
+                        .verified_certs
+                        .read()
+                        .unwrap()
+                        .contains(&cert.cert_type)
+                    {
+                        self.stats.received_verified.fetch_add(1, Ordering::Relaxed);
+                        packet.meta_mut().set_discard(true);
+                        continue;
+                    }
+
+                    certs_to_verify.push(cert);
+                }
+            }
+        }
+        (votes_to_verify, certs_to_verify)
+    }
+
+    pub fn verify_and_send_batches(
+        &mut self,
+        mut batches: Vec<PacketBatch>,
+        pool: Option<&ThreadPool>,
+    ) -> Result<(), BLSSigVerifyServiceError> {
+        let mut verify_time = Measure::start("sigverify_batch_time");
+        let mut preprocess_time = Measure::start("preprocess");
+        let mut consensus_metrics_to_send = Vec::with_capacity(AVERAGE_VOTES_OR_CERTS_PER_BATCH);
+        let root_bank = self.sharable_banks.root();
+
+        self.purge_old_votes_and_certs(root_bank.slot());
+
+        preprocess_time.stop();
+        self.stats.preprocess_count.fetch_add(1, Ordering::Relaxed);
+        self.stats
+            .preprocess_elapsed_us
+            .fetch_add(preprocess_time.as_us(), Ordering::Relaxed);
+
+        let (votes_to_verify, certs_to_verify) =
+            self.preprocess_packets(&mut batches, &root_bank, &mut consensus_metrics_to_send);
+
+        let (votes_result, certs_result) = Self::run_batch_verify_in_pool(pool, || {
+            rayon::join(
+                || self.verify_and_send_votes(votes_to_verify),
+                || self.verify_and_send_certificates(certs_to_verify, &root_bank),
+            )
+        });
+        verify_time.stop();
+        self.stats
+            .verify_elapsed_us
+            .fetch_add(verify_time.as_us(), Ordering::Relaxed);
+
+        votes_result?;
+        certs_result?;
+
+        // Send to metrics service for metrics aggregation
+        if self
+            .consensus_metrics_sender
+            .send((Instant::now(), consensus_metrics_to_send))
+            .is_err()
+        {
+            warn!("could not send consensus metrics, receive side of channel is closed");
+        }
```

Comment:
I thought if the channel is disconnected, that probably means we are shutting down so we should propagate an error back?